File: 0907-koko-eating-bananas/solution.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int n= piles.length;
        //base case
        if(n > h) return -1;
        
        int maxi=piles[0];
        for(int i=0; i<n; i++){
            maxi = Math.max( maxi, piles[i]);
        }
        
        int low=1, high = maxi, ans = -1;
        while( low <= high){
            int mid = low + (high - low)/2;
            
            if( isPossibleWithKbanana(mid, piles, h)){
                ans = mid;
                high= mid-1;
            }
            else{
                low = mid+1;
            }
        }
        return ans;
    }
    public boolean isPossibleWithKbanana(int capacity, int []piles, int h){
        int time =0;
        for(int i=0; i<piles.length; i++){
            if( piles[i] % capacity == 0)
                time  = time + piles[i]/capacity;
            else{
                time = time + ( piles[i]/capacity ) +1;
            }
            
            if(time > h) return false;
        }
        if( time <= h) return true;
        return false;
    }
}

File: 0907-koko-eating-bananas/solution.js
/**
 * @param {number[]} piles
 * @param {number} h
 * @return {number}
 */
var minEatingSpeed = function(piles, h) {
    let maxi =1;
    for(let x of piles){
        if(x > maxi) maxi = x;
    }

    let s=1, e=maxi;
    let ans=maxi;
    while(s<=e){
        let mid = s + Math.floor( (e-s)/2 );

        let result = isPossibleWithMidSpeed(mid, piles, h);

        if(result == 1) {
            ans = mid;
            e= mid-1;
        }
        else s=mid+1;
    }
    return  ans;
};

var isPossibleWithMidSpeed = function(mid, piles, h){
    let timeTaken = 0;
    for(let x of piles){
        let time = Math.ceil(x/mid);
        timeTaken +=time;
        if(timeTaken > h) return 2; // we have exceed max allowed time
    }
    return 1;  // here totalTime is either less than n or equal to n
}
